Upgrade to OpenLiteSpeed 1.7.18

## OpenLiteSpeed v1.7.18
In this release: security updates, and more!
### RELEASE LOG:
- [Security] Apply more strict request header validation.
- [Security] Update libmodsecurity to 3.10
- [Tuning] Lift default memory limit for external application.
- [Improvement] add private cache session cookies detection for WordPress. 

https://openlitespeed.org/release-log/

```
configure: SSDEEP library was not found
*** LOOKING AT PATH:  /usr/lib
*** LOOKING AT PATH:  /usr/local/lib
*** LOOKING AT PATH:  /usr/local/lib64
*** LOOKING AT PATH:  /usr/local/lua
*** LOOKING AT PATH:  /usr/local/liblua
*** LOOKING AT PATH:  /usr/local
*** LOOKING AT PATH:  /opt
*** LOOKING AT PATH:  /usr
*** LOOKING AT PATH:  /usr/lib64
*** LOOKING AT PATH:  /opt/local
*** LOOKING AT PATH:  /usr/lib/lua5.3/liblua
*** LOOKING AT PATH:  /usr/lib/lua5.2/liblua
configure: LUA library was not found
checking for libcurl config script... no
configure: *** curl library not found.
configure: error: Curl was explicitly referenced but it was not found
make: *** No targets specified and no makefile found.  Stop.
cp: cannot stat 'src/.libs/libmodsecurity.a': No such file or directory
```

```
/usr/bin/ld: cannot find -ldecrepit
collect2: error: ld returned 1 exit status
make[2]: *** [src/CMakeFiles/openlitespeed.dir/build.make:159: src/openlitespeed] Error 1
make[1]: *** [CMakeFiles/Makefile2:829: src/CMakeFiles/openlitespeed.dir/all] Error 2
make: *** [Makefile:149: all] Error 2
cp: cannot stat 'build/src/openlitespeed': No such file or directory
```

--- a/.github/workflows/docker.yml
+++ b/.github/workflows/docker.yml
@@ -1,4 +1,4 @@
-name: docker-build
+name: release-build
 
 on:
   workflow_dispatch:
@@ -7,8 +7,8 @@ on:
       - main
 
 env:
-  OLS_VERSION: 1.7.17
-  PHP_STABLE_VERSION: '8.2.8'
+  OLS_VERSION: 1.7.18
+  PHP_STABLE_VERSION: '8.2.9'
   REGISTRY: ghcr.io
 
 jobs:
@@ -17,7 +17,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix: 
-        PHP_VERSION: ['8.0.29', '8.1.21', '8.2.8']
+        PHP_VERSION: ['8.0.30', '8.1.22', '8.2.9']
 
     steps:
       - name: Checkout

--- a/.github/workflows/test-builds.yml
+++ b/.github/workflows/test-builds.yml
@@ -1,0 +1,67 @@
+name: docker-build
+
+on:
+  workflow_dispatch:
+  pull_request:
+    branches:
+      - develop
+
+env:
+  OLS_VERSION: 1.7.17
+  PHP_STABLE_VERSION: '8.2.8'
+  REGISTRY: ghcr.io
+
+jobs:
+  buildx:
+    runs-on: self-hosted
+    strategy:
+      fail-fast: false
+      matrix: 
+        PHP_VERSION: ['8.0.29', '8.1.21', '8.2.8']
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0 # Required due to the way Git works, without it this action won't be able to find any or the correct tags
+
+      # https://github.com/marketplace/actions/docker-setup-buildx
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+
+      # https://github.com/marketplace/actions/docker-login
+      - name: Login to GitHub Packages
+        uses: docker/login-action@v2
+        with:
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+          logout: false
+
+      # https://github.com/orgs/community/discussions/26625#discussioncomment-3252582
+      - name: Determine PHP Major/Minor Version
+        id: php-version
+        run: |
+          _0=$(echo ${{ matrix.PHP_VERSION }} | cut -d. -f1)
+          _1=$(echo ${{ matrix.PHP_VERSION }} | cut -d. -f2)
+          echo "_0=$_0" >> $GITHUB_OUTPUT
+          echo "_1=$_1" >> $GITHUB_OUTPUT
+
+      # https://github.com/marketplace/actions/build-and-push-docker-images
+      - name: Build Docker Images
+        uses: docker/build-push-action@v4
+        with:
+          context: template
+          platforms: linux/arm64
+          provenance: false
+          build-args: |
+            OLS_VERSION=${{ env.OLS_VERSION }}
+            PHP_VERSION=${{ matrix.PHP_VERSION }}
+            PHP_MAJOR_VERSION=${{ steps.php-version.outputs._0 }}
+            PHP_MINOR_VERSION=${{ steps.php-version.outputs._1 }}
+          push: false
+          tags: ${{ env.REGISTRY }}/ndigitals/openlitespeed:${{ env.OLS_VERSION }}-lsphp${{ steps.php-version.outputs._0 }}${{ steps.php-version.outputs._1 }}
+          no-cache: ${{ github.event_name == 'workflow_dispatch' && true || false }}
+          cache-from: type=registry,ref=${{ env.REGISTRY }}/ndigitals/openlitespeed:latest
+          cache-to: type=inline
+

--- a/.github/workflows/test-builds.yml
+++ b/.github/workflows/test-builds.yml
@@ -1,4 +1,4 @@
-name: docker-build
+name: test-build
 
 on:
   workflow_dispatch:
@@ -7,8 +7,8 @@ on:
       - develop
 
 env:
-  OLS_VERSION: 1.7.17
-  PHP_STABLE_VERSION: '8.2.8'
+  OLS_VERSION: 1.7.18
+  PHP_STABLE_VERSION: '8.2.9'
   REGISTRY: ghcr.io
 
 jobs:
@@ -17,7 +17,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix: 
-        PHP_VERSION: ['8.0.29', '8.1.21', '8.2.8']
+        PHP_VERSION: ['8.0.30','8.1.22','8.2.9']
 
     steps:
       - name: Checkout
@@ -48,7 +48,7 @@ jobs:
           echo "_1=$_1" >> $GITHUB_OUTPUT
 
       # https://github.com/marketplace/actions/build-and-push-docker-images
-      - name: Build Docker Images
+      - name: Build Docker Image
         uses: docker/build-push-action@v4
         with:
           context: template
@@ -59,9 +59,30 @@ jobs:
             PHP_VERSION=${{ matrix.PHP_VERSION }}
             PHP_MAJOR_VERSION=${{ steps.php-version.outputs._0 }}
             PHP_MINOR_VERSION=${{ steps.php-version.outputs._1 }}
+          load: true
           push: false
-          tags: ${{ env.REGISTRY }}/ndigitals/openlitespeed:${{ env.OLS_VERSION }}-lsphp${{ steps.php-version.outputs._0 }}${{ steps.php-version.outputs._1 }}
+          tags: openlitespeed:${{ env.OLS_VERSION }}-lsphp${{ steps.php-version.outputs._0 }}${{ steps.php-version.outputs._1 }}
           no-cache: ${{ github.event_name == 'workflow_dispatch' && true || false }}
           cache-from: type=registry,ref=${{ env.REGISTRY }}/ndigitals/openlitespeed:latest
           cache-to: type=inline
+
+      - name: Test Docker Image
+        run: |
+          IMAGE=openlitespeed:${{ env.OLS_VERSION }}-lsphp${{ steps.php-version.outputs._0 }}${{ steps.php-version.outputs._1 }}
+          ID=$(docker run -d ${IMAGE})
+          sleep 5s
+          docker exec -i ${ID} su -c 'mkdir -p /var/www/vhosts/localhost/html/ && echo "<?php phpinfo();" > /var/www/vhosts/localhost/html/index.php && service lsws restart'
+          HTTP=$(docker exec -i ${ID} su -c 'curl -s -o /dev/null -Ik -w "%{http_code}" http://localhost')
+          HTTPS=$(docker exec -i ${ID} su -c 'curl -s -o /dev/null -Ik -w "%{http_code}" https://localhost')
+          docker kill ${ID} &>/dev/null
+          docker rm ${ID} &>/dev/null
+          if [[ "${HTTP}" != "200" || "${HTTPS}" != "200" ]]; then
+              echo -e '[\u2718] Test failed!'
+              echo "http://localhost returned ${HTTP}"
+              echo "https://localhost returned ${HTTPS}"
+              exit 1
+          else
+              docker rmi ${IMAGE} &>/dev/null
+              echo -e '[\u2714] Tests passed!'
+          fi
 

--- a/template/Dockerfile
+++ b/template/Dockerfile
@@ -40,15 +40,13 @@ COPY *.sh /build/
 #     - https://www.dajobe.org/blog/2015/04/18/making-debian-docker-images-smaller/
 #     - https://github.com/dajobe/docker-nghttp2
 RUN /build/secure-base.sh && \
+    /build/prepare-runtime.sh && \
+    /build/prepare-ols.sh && \
+    echo 'cloud-docker' > /usr/local/lsws/PLAT && \
     /build/prepare-build.sh && \
     /build/prepare-php.sh && \
     cd /build/php-$PHP_VERSION && make -j3 && make install && \
     /build/config-php.sh && \
-    /build/prepare-ols.sh && \
-    cd /build/openlitespeed-$OLS_VERSION && ./build.sh && ./install.sh && \
-    echo 'cloud-docker' > /usr/local/lsws/PLAT && \
-    cd /build/openlitespeed-$OLS_VERSION/src/modules/modsecurity-ls && make -f Makefile.f && \
-    cp mod_security.so /usr/local/lsws/modules && \
     /build/prepare-ols-admin_php.sh && \
     cd /usr/local/lsws/phpbuild/php-$OLS_ADMIN_PHP_VERSION && make -j3 && \
     /build/config-ols-admin_php.sh && \

--- a/template/cleanup-build-env.sh
+++ b/template/cleanup-build-env.sh
@@ -12,9 +12,6 @@ rm -f "$tmp"
 
 apt-get remove --purge -y $PACKAGES_TO_REMOVE
 
-# Install the run-time dependencies
-apt-get install $MINIMAL_APT_GET_ARGS $RUN_PACKAGES
-
 rm -rf /tmp/* /var/tmp/* $LSWS_HOME/phpbuild
 
 apt-get clean

--- a/template/config-build-env.sh
+++ b/template/config-build-env.sh
@@ -11,23 +11,22 @@ MINIMAL_APT_GET_ARGS='-y --no-install-recommends'
 # File containing original installed packages
 PACKAGES_INSTALLED_LOG="/tmp/packages.lst"
 
+# Initial Setup Packages.
+SETUP_PACKAGES="ca-certificates cron tzdata pkg-config git curl wget"
+
 ## Run time dependencies ##
-RUN_PACKAGES="ca-certificates cron tzdata openssl mariadb-client libgssapi-krb5-2 libkrb5-3 libexpat1 libxml2 libargon2-1 libenchant-2-2 libpng16-16 libwebp6 libjpeg62-turbo libxpm4 libfreetype6 libonig5 libsodium23 libxslt1.1 libzip4 libzstd1 liblz4-1 libcurl4 imagemagick libc-client2007e libmemcached11 libdbd-freetds freetds-bin procps libatomic1 net-tools less libjpeg-turbo-progs optipng gifsicle zip unzip libyajl2 libpcre2-posix2 libpcre++0v5 liblmdb0 libgeoip1 ruby-full gnupg2 nodejs yarn"
+RUN_PACKAGES="ca-certificates cron tzdata pkg-config git curl wget openssl mariadb-client libgssapi-krb5-2 libkrb5-3 libexpat1 libxml2 libargon2-1 libenchant-2-2 libpng16-16 libwebp6 libjpeg62-turbo libxpm4 libfreetype6 libonig5 libsodium23 libxslt1.1 libzip4 libzstd1 liblz4-1 libcurl4 imagemagick libc-client2007e libmemcached11 libdbd-freetds freetds-bin procps libatomic1 net-tools less libjpeg-turbo-progs optipng gifsicle zip unzip libyajl2 libpcre2-posix2 libpcre++0v5 liblmdb0 libgeoip1 ruby-full gnupg2 nodejs yarn mmdb-bin lua-mmdb brotli libbrotli1 liblua5.2-0 lua5.2 ssdeep libldap-2.4-2 libldap-common libssh2-1 libsasl2-2 libsasl2-modules libsasl2-modules-db libnghttp2-14 libpsl5 librtmp1 publicsuffix"
 
 ## Build time dependencies ##
 
-# git is needed for git clone; not building
-# alternate would be to download a release tarball with curl or wget
-BUILD_PACKAGES="git curl wget"
-
 # Core list from docs
-BUILD_PACKAGES="$BUILD_PACKAGES pkg-config"
+BUILD_PACKAGES="build-essential"
 
 # OLS build required packages
-BUILD_PACKAGES="$BUILD_PACKAGES cmake g++ libgeoip-dev liblmdb-dev libyajl-dev libpcre++-dev"
+BUILD_PACKAGES="$BUILD_PACKAGES cmake g++ autoconf automake libtool flex bison gdb libgeoip-dev liblmdb-dev libyajl-dev libpcre++-dev libmaxminddb-dev liblualib50-dev liblua50-dev libudns-dev libbrotli-dev libssl-dev zlib1g-dev liblua5.2-dev libcurl4-openssl-dev libexpat1-dev libcurlpp-dev libcurlpp0 libidn11-dev libkrb5-dev libldap2-dev librtmp-dev libssh2-1-dev"
 
 # PHP building required packages
-BUILD_PACKAGES="$BUILD_PACKAGES build-essential libssl-dev libdb5.3-dev krb5-multidev libkrb5-dev bison autoconf automake libtool re2c flex libxml2-dev libbz2-dev libcurl4-openssl-dev libexpat1-dev libjpeg-dev libfreetype6-dev libgmp3-dev libpng-dev libxpm-dev libc-client2007e-dev libenchant-2-dev libsasl2-dev libc-client2007e-dev libldap2-dev libldb-dev libmcrypt-dev libmhash-dev freetds-dev zlib1g-dev libpq-dev libmariadb-dev-compat libmariadb-dev libncurses5-dev libpcre2-dev libpcre3-dev unixodbc-dev libsqlite3-dev libaspell-dev libreadline6-dev librecode-dev libsnmp-dev libtidy-dev libxslt-dev libonig-dev libzip-dev libwebp-dev freetds-dev libpspell-dev libedit-dev libsodium-dev libargon2-dev libvarnishapi-dev libmagickwand-dev libmagickcore-dev libmemcached-dev libzstd-dev liblz4-dev libyaml-dev libffi-dev"
+BUILD_PACKAGES="$BUILD_PACKAGES libdb5.3-dev krb5-multidev re2c libxml2-dev libbz2-dev libjpeg-dev libfreetype6-dev libgmp3-dev libpng-dev libxpm-dev libc-client2007e-dev libenchant-2-dev libsasl2-dev libc-client2007e-dev libldb-dev libmcrypt-dev libmhash-dev freetds-dev zlib1g-dev libpq-dev libmariadb-dev-compat libmariadb-dev libncurses5-dev libpcre2-dev libpcre3-dev unixodbc-dev libsqlite3-dev libaspell-dev libreadline6-dev librecode-dev libsnmp-dev libtidy-dev libxslt-dev libonig-dev libzip-dev libwebp-dev freetds-dev libpspell-dev libedit-dev libsodium-dev libargon2-dev libvarnishapi-dev libmagickwand-dev libmagickcore-dev libmemcached-dev libzstd-dev liblz4-dev libyaml-dev libffi-dev"
 
 # apt-get remove --allow-remove-essential enters an infinite loop of
 # pam errors with this package

--- a/template/config-ols.sh
+++ b/template/config-ols.sh
@@ -4,11 +4,6 @@
 
 cd /build || exit
 
-wget -O $LSWS_HOME/admin/misc/lsup.sh \
-    https://raw.githubusercontent.com/litespeedtech/openlitespeed/master/dist/admin/misc/lsup.sh
-
-chmod +x $LSWS_HOME/admin/misc/lsup.sh
-
 echo "listener HTTP {
   address                 *:80
   secure                  0

--- a/template/prepare-build.sh
+++ b/template/prepare-build.sh
@@ -3,28 +3,5 @@
 . /build/config-build-env.sh
 
 apt-get update -y
-apt install curl -y
-
-# Prepare to install Node 16.
-curl -sL https://deb.nodesource.com/setup_16.x | bash -
-
-# Prepare to install Yarn.
-curl -sL https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add -
-echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list
-
-apt-get update -y
-
-dpkg --get-selections | awk '{print $1}' | sort > "$PACKAGES_INSTALLED_LOG"
-echo "Installed packages at start"
-cat "$PACKAGES_INSTALLED_LOG"
-
-apt-get install $MINIMAL_APT_GET_ARGS $BUILD_PACKAGES $RUN_PACKAGES
-
-# https://github.com/SpiderLabs/ModSecurity/issues/804#issuecomment-246158487
-#cd /build || exit
-
-# Download and prepare YAJL source.
-#wget https://github.com/lloyd/yajl/archive/refs/tags/2.1.0.tar.gz -O yajl-2.1.0.tar.gz && \
-#	tar xzf yajl-2.1.0.tar.gz && cd yajl-2.1.0 && \
-#	./configure
+apt-get install $MINIMAL_APT_GET_ARGS $BUILD_PACKAGES
 

--- a/template/prepare-ols.sh
+++ b/template/prepare-ols.sh
@@ -7,7 +7,17 @@ cd /build || exit
 # carefully, it's import to export PKG_CONFIG_PATH to tell the pkg-config the info about yajl
 export PKG_CONFIG_PATH=/usr/local/share/pkgconfig:PKG_CONFIG_PATH
 
-# Download and prepare OpenLiteSpeed source.
-wget https://openlitespeed.org/packages/openlitespeed-$OLS_VERSION.src.tgz && \
-	tar xzf openlitespeed-$OLS_VERSION.src.tgz
+# Configure OpenLiteSpeed binary repository.
+REPO_OS_CODENAME=$(cat /etc/os-release |  grep '^VERSION_CODENAME=' | head -n1 | awk -F '=' '{print $2}' | tr -d '"')
+
+wget -O /etc/apt/trusted.gpg.d/lst_debian_repo.gpg http://rpms.litespeedtech.com/debian/lst_debian_repo.gpg && \
+  wget -O /etc/apt/trusted.gpg.d/lst_repo.gpg http://rpms.litespeedtech.com/debian/lst_repo.gpg && \
+  echo "deb http://rpms.litespeedtech.com/debian/ ${REPO_OS_CODENAME} main" > /etc/apt/sources.list.d/lst_debian_repo.list && \
+  echo "#deb http://rpms.litespeedtech.com/edge/debian/ ${REPO_OS_CODENAME} main" >> /etc/apt/sources.list.d/lst_debian_repo.list && \
+  apt-get update -y && \
+  apt-get install $MINIMAL_APT_GET_ARGS openlitespeed=${OLS_VERSION}* ols-modsecurity
+
+dpkg --get-selections | awk '{print $1}' | sort > "$PACKAGES_INSTALLED_LOG"
+echo "Installed packages at start"
+cat "$PACKAGES_INSTALLED_LOG"
 

--- a/template/prepare-runtime.sh
+++ b/template/prepare-runtime.sh
@@ -1,0 +1,17 @@
+#!/bin/bash -x
+
+. /build/config-build-env.sh
+
+apt-get update -y
+apt-get install $MINIMAL_APT_GET_ARGS $SETUP_PACKAGES
+
+# Prepare to install Node 16.
+curl -sL https://deb.nodesource.com/setup_16.x | bash -
+
+# Prepare to install Yarn.
+curl -sL https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add -
+echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list
+
+apt-get update -y
+apt-get install $MINIMAL_APT_GET_ARGS $RUN_PACKAGES
+

